<img src="IMAGES_PYTHON_COURSE_2018/LOGO_MEINBIO_TRANSPARENT.png" width="200" height="200" />

# <center>Python course 2018
### <center>MeInBio Training Group 
#### <center> by Florian Heyl & Francesco Ferrari 

# PART 5: CONDITIONAL STATEMENTS

An interesting piece of program is one that can evaluate if certain conditions are met and act accordingly. For example, imagine we have a DNA sequence and we want to print it to the screen if the sequence is longer than 300 bp. Or imagine we have a count table (from an RNA-Seq experiment) with the number of reads mapped to each gene and we want our program to discard a gene if that gene has less that 10 reads mapped to it. 

Whenever we want our program to verify if a condition is met, we use a conditional statement. Let's have a look at what are conditions and how conditional statements work in Python. 

### What is a condition? 

A condition is simply a bit of code that can produce a true or false answer.

In [1]:
# examples of conditions

print(len('ATCCGTA') > 2)
print(20 == 30) ### PLEASE NOTICE THAT WE USE "==" AND NOT "=" TO TEST IF TWO ITEMS ARE THE SAME
print("I Love Pizza".startswith('I'))
print("I Also Love Carbonara".endswith("y"))
print("apple" in ["banana","orange","cherry"])
print("Anakin" != "Luke")

True
False
True
False
False
True


- # If else Statements

We have just seen how to build some conditions and have our code to output True or False if the condition is met or not met respectively. Now, we are ready to make our program take decisions and act according to a condition's outcome.

The simplest kind of conditional statement is an "if statement". In Python, the syntax for if statements in very easy:

We write the word "if", followed by a condition, and end the first line with a colon. There follows a block of indented lines of code (the body of the if statement), which will only be executed if the condition is true.

In [2]:
# example of if statement 1 (easy)

#my_sequence = "AAAATCCCCGTACT"
my_sequence = "AAT"
if len(my_sequence) > 4:
    print("my sequence is long".upper())
else:
    print("my sEQUence Is short".lower())

my sequence is short


### Exercise n° 1 (medium): 

Given a list of tuples of the form (chr, gene_name), identify all genes present in the X chromosome and save them to a text file.

In [3]:
import sys
import os
sys.path.append(os.path.abspath("./Book_and_other_resources/"))

In [4]:
import python_course_functions as pcf
gene_list = pcf.loadGeneList("./BOOK_and_other_resources/gene_list.txt")

print("You are working with {} genes".format(len(gene_list)))

gene_list[0:10] # to see the structure of your data

You are working with 53715 genes


[('chr1', 'ENSMUSG00000102693.1'),
 ('chr1', 'ENSMUSG00000064842.1'),
 ('chr1', 'ENSMUSG00000051951.5'),
 ('chr1', 'ENSMUSG00000102851.1'),
 ('chr1', 'ENSMUSG00000103377.1'),
 ('chr1', 'ENSMUSG00000104017.1'),
 ('chr1', 'ENSMUSG00000103025.1'),
 ('chr1', 'ENSMUSG00000089699.1'),
 ('chr1', 'ENSMUSG00000103201.1'),
 ('chr1', 'ENSMUSG00000103147.1')]

In [5]:
print(gene_list[0][0] == "chr1")
#len(gene_list[0])
for char in "mystring":
    print(char)

True
m
y
s
t
r
i
n
g


In [9]:
### insert your commands in the space below

X_chrom = open('test_data/X_chrom.csv', 'w')
for file in gene_list:
    if file[0] =="chrX":
        gene = file[1] + "\n"
        X_chrom.write(gene)
        
X_chrom.close()


# 2nd solution
newfile = open('test_data/allXchromgenes.csv', 'w')
for i in range(0, len(gene_list)):
    file = gene_list[i]
    chromosome = gene_list[i][0]
    if(chromosome == "chrX"):
        newfile.write(gene_list[i][0] + gene_list[i][1] + "\n")

In [10]:
# 3rd solution
my_file = "test_data/chrx_genes.csv"

file = open("test_data/chrx_genes.csv", "w")

def write_chrx_genes(x, file):
    
    file = open(file, "w")

    for i in range(0, len(x)):
        chromosome = x[i][0]
        if(chromosome == "chrX"):
            file.write(x[i][0] + x[i][1] + "\n")

    file.close()

### needs to work:
write_chrx_genes(gene_list, "test_data/chrx_genes.csv")

- ##  Elif statements

If we have more than 2 possible branches, instead of using else, we use elif. 

In [11]:
# example of elif statement
my_fav_fruit=["banana", "apple", "cherry", "kiwi", "pear", "rasberry", "blueberry", "peach", "cas"]

for fruit in my_fav_fruit:
    if len(fruit) == 4:
        print("'{}' has 4 letters\n".format(fruit))
    elif len(fruit) == 5:
        print("'{}' has 5 letters\n".format(fruit))
    elif len(fruit) > 5:
        print("'{}' is a very long word\n".format(fruit))
    else:
        print("'{}' is a very short word\n".format(fruit))

'banana' is a very long word

'apple' has 5 letters

'cherry' is a very long word

'kiwi' has 4 letters

'pear' has 4 letters

'rasberry' is a very long word

'blueberry' is a very long word

'peach' has 5 letters

'cas' is a very short word



### Exercise n° 2 (medium): 

Using the same list of tuples of the form (chr, gene_name) from exercise 1, write to separate files all genes found on chromosome 1, chromosome 4, chromosome 11 and chromosome Y

In [12]:
### insert your commands in the space below

# gene_list = pcf.loadGeneList("./BOOK_and_other_resources/gene_list.txt") has to be executed before!
gene_list = pcf.loadGeneList("./BOOK_and_other_resources/gene_list.txt")

chrom_X = open('test_data/chrom_X.csv',"w")
chrom_1 = open('test_data/chrom_1.csv',"w")
chrom_4 = open('test_data/chrom_4.csv',"w")
chrom_11 = open('test_data/chrom_11.csv',"w")
chrom_Y = open('test_data/chrom_Y.csv',"w")

for file in gene_list:
    if file[0] =="chrX":
        gene = file[1] + "\n"
        chrom_X.write(gene)
    elif file[0] =="chr1":
        gene = file[1] + "\n"
        chrom_1.write(gene)
    elif file[0] =="chr4":
        gene = file[1] + "\n"
        chrom_4.write(gene)
    elif file[0] =="chr11":
        gene = file[1] + "\n"
        chrom_11.write(gene)
    elif file[0] =="chrY":
        gene = file[1] + "\n"
        chrom_Y.write(gene)
        
chrom_X.close()
chrom_1.close()
chrom_4.close()
chrom_11.close()
chrom_Y.close()

- # Combining Conditions with and / or

It is possible to express conditions made up of several parts. For example, given a list of accession IDs, print all IDs that start with "a"/"A" and end with "z/"Z". To do that, we use the operator "and". The combined "if" condition will return a True if all conditions are satisfied, False otherwise. 

Alternatively, we can write disjunctive conditions using the "or" logical operator. In this case, the combined "if" condition will return a True if at least one of the conditions are satisfied, False if all condition are not satisfied.    

In [13]:
list_ID = ["ID_a132i398zzz",
           "ID_b34573y4jnj",
           "ID_S45u92s8fgz",
           "ID_Afkj3ehrdZZ",
           "ID_Askdsa3s4nz",
           "ID_EFGJNG24OIN",
           "ID_234oitfunkj",
           "ID_ASOI0934U5Z",
           "ID_ewowrit13ir"]

print("IDs starting with 'a' AND ending with 'z':")
for i in list_ID:
    id_seq = i.split("_")[1]
    if id_seq[0].lower() == "a" and id_seq[-1].lower() == "z":
        print(i)
print("\n")     
print("IDs starting with 'e' OR ending with 'j':")
for i in list_ID:
    id_seq = i.split("_")[1]
    if id_seq[0].lower() == "e" or id_seq[-1].lower() == "j":
        print(i)

IDs starting with 'a' AND ending with 'z':
ID_a132i398zzz
ID_Afkj3ehrdZZ
ID_Askdsa3s4nz
ID_ASOI0934U5Z


IDs starting with 'e' OR ending with 'j':
ID_b34573y4jnj
ID_EFGJNG24OIN
ID_234oitfunkj
ID_ewowrit13ir


In [14]:
list_ID = ["ID_a132i398zzz",
           "ID_b34573y4jnj",
           "ID_S45u92s8fgz",
           "ID_Afkj3ehrdZZ",
           "ID_Askdsa3s4nz",
           "ID_EFGJNG24OIN",
           "ID_234oitfunkj",
           "ID_ASOI0934U5Z",
           "ID_ewowrit13ir"]

i = list_ID[0]
print(i)
print(type(i.split("_")[0]))


ID_a132i398zzz
<class 'str'>


### Exercise n° 3 (medium):

A) Using the gene list from exercise 1, count how many genes are found in chromosome 1 and have a ID that ends with 1. 

B) Using the same list, count how many genes are found in chromosome 2 and 4 which have and ID whose number after the dot is either a 3 or a 7. (PAY ATTENTION! THIS IS DIFFERENT FROM THE CONDITION OF 3A).  

In [15]:
### insert your commands in the space below
gene_list = pcf.loadGeneList("./BOOK_and_other_resources/gene_list.txt")

# A) 
count = []
counter = 0
for file in gene_list:
    if file[0] == "chr1" and file[1][-1] == "1":
        count.append(file[1])
        counter += 1

print(len(count))
print(counter)

2092
2092


In [16]:
# B) 
# chromosome 2 and 4 which have and ID whose number after the dot is either a 3 or a 7
counter = 0
for file in gene_list:
    ID = file[1].split(".")
    if (file[0] == "chr2" or file[0] == "chr4") and (ID[1][0] == "3" or ID[1][0] == "7"):
        counter += 1
print(counter)

559


Expected Outcome:


3A): 2092          
3B): 559

- # Boolean functions and conditional statements

Sometime, you may need to check if a complex condition (that means, a condition that cannot be expressed with the standard tools "==", ">", "<", "endswith", "startswith", etc.) holds true and have your script react according to the outcome.

To do that, we must write our own function that returns a True if the condition we want to test is met, and False otherwise. 

For example:
  
Say that you want to check if a sequence is AT-rich. If the sequence is AT-rich, then you print it to the screen with the corresponding percentage. You also want to collect all not AT-rich regions in a list, and print it to the screen.

In [17]:
def is_AT_rich(sequence):
    sequence = sequence.lower()  # sequence is made of lower characters to avoid mismatches due to capitalization
    a = sequence.count("a") # count how many "a" are in the sequence
    t = sequence.count("t") # count how many "t" are in the sequence
    fr_at = (a+t)/len(sequence) # compute the fraction of at present in the sequence
    if fr_at >= 0.65:    # if the fraction is greater or equal 0.65 (65%) ...
        return (True, fr_at) # return a tuple where the first element is a True (boolean), and the second is the at fraction
    else:                   # otherwise
        return (False, fr_at)  # return a tuple where the first element is a False (boolean), and the second is the at fraction 
    
sequences = ["ATTTGTCCAAAGTATATAT",
             "GGGCGAGGCGTTGCGAGGG",
             "GGCGAAAAAAATTTTTTTT",
             "AAATATATATATATATATA",
             "GGATTATCCTGGATATATAT",
             "TTTTTTTTTTTAAAAAAAAA",
             "GGGTTTTTATATATTTAAAT",
             "GGGTGGGCGCGCGCGAGGGG"]

not_AT_rich = []
for sequence in sequences:
    if is_AT_rich(sequence)[0]:
        print("{} is composed for the {}% of AT\n".format(sequence, round(is_AT_rich(sequence)[1]*100,1) ))
    else:
        not_AT_rich.append(sequence)
print("{} sequence were not AT-rich:{}".format(len(not_AT_rich), not_AT_rich))

ATTTGTCCAAAGTATATAT is composed for the 78.9% of AT

GGCGAAAAAAATTTTTTTT is composed for the 78.9% of AT

AAATATATATATATATATA is composed for the 100.0% of AT

GGATTATCCTGGATATATAT is composed for the 70.0% of AT

TTTTTTTTTTTAAAAAAAAA is composed for the 100.0% of AT

GGGTTTTTATATATTTAAAT is composed for the 85.0% of AT

2 sequence were not AT-rich:['GGGCGAGGCGTTGCGAGGG', 'GGGTGGGCGCGCGCGAGGGG']


### Exercise n° 4 (difficult):

write a function that verifies whether a in a sequence there are at least 4 CpG (a C immediately followed by a G, i.e. "CG") and return a True if so, False otherwise. 

Then, given a list of sequence, print to screen those sequences that have at least 4 CpG in them.

In [19]:
new_sequences = ["CGGGGAGTCGAAATTGCTA",
                 "ATTCGCGCGCGATAATTCG",
                 "TTTCGGCCGCGATTTACGG",
                 "TTAATTCGGGATCGCGAAT"]

### insert your commands in the space below
def isCpGrich(sequence):
    numCpG = sequence.count("CG")
    if numCpG >= 4:
        return True
    else:
        return False

for sequence in new_sequences:
    if isCpGrich(sequence):
        print(sequence)


ATTCGCGCGCGATAATTCG
TTTCGGCCGCGATTTACGG


Expected Outcome:

ATTCGCGCGCGATAATTCG

TTTCGGCCGCGATTTACGG